src/hotspot/share/gc/shenandoah/shenandoahHeap.inline.hpp

```diff
@@ -488,6 +488,30 @@ inline void ShenandoahHeap::mark_complete_marking_context() {
   _marking_context->mark_complete();
 }
 
+inline void ShenandoahHeap::mark_finished() {
+  mark_complete_marking_context();
+
+  class ShenandoahCollectLiveSizeClosure : public ShenandoahHeapRegionClosure {
+  private:
+    size_t _live;
+
+  public:
+    ShenandoahCollectLiveSizeClosure() : _live(0) {}
+
+    void heap_region_do(ShenandoahHeapRegion* r) {
+      _live += r->get_live_data_bytes();
+    }
+
+    size_t get_live() {
+      return _live;
+    }
+  };
+
+  ShenandoahCollectLiveSizeClosure cl;
+  heap_region_iterate(&cl);
```

Comment:
I think you want `parallel_heap_region_iterate` on this path, and do `Atomic::add(&_live, r->get_live_data_bytes())` in the closure. We shall see if this makes sense to make fully concurrently...

Reply:
Outdated by the patch provided by @shipilev